module/VuFind/src/VuFind/Auth/LoginToken.php

```diff
@@ -0,0 +1,280 @@
+        \Laminas\Config\Config $config,
+        \VuFind\Db\Table\User $userTable,
+        \VuFind\Db\Table\LoginToken $loginTokenTable,
+        \VuFind\Cookie\CookieManager $cookieManager,
+        SessionManager $sessionManager,
+        \VuFind\Mailer\Mailer $mailer,
+    ) {
+        $this->config = $config;
+        $this->userTable = $userTable;
+        $this->loginTokenTable = $loginTokenTable;
+        $this->cookieManager = $cookieManager;
+        $this->sessionManager = $sessionManager;
+        $this->mailer = $mailer;
+    }
+
+    /**
+     * Authenticate user using a login token cookie
+     *
+     * @param string $sessionId Session identifier
+     *
+     * @return \VuFind\Db\Row\UserRow Object representing logged-in user.
+     */
+    public function tokenLogin(string $sessionId): ?\VuFind\Db\Row\User
+    {
+        $cookie = $this->getLoginTokenCookie();
+        $user = null;
+        if ($cookie) {
+            try {
+                if ($token = $this->loginTokenTable->matchToken($cookie)) {
+                    $this->loginTokenTable->deleteBySeries($token->series, $token->user_id);
+                    $user = $this->userTable->getById($token->user_id);
+                    $this->createToken($user, $token->series, $sessionId, $token->expires);
+                }
+            } catch (LoginTokenException $e) {
+                // Delete all login tokens for the user and all sessions
+                // associated with the tokens and send a warning email to user
+                $user = $this->userTable->getById($cookie['user_id']);
+                $userTokens = $this->loginTokenTable->getByUserId($cookie['user_id']);
+                $handler = $this->sessionManager->getSaveHandler();
+                foreach ($userTokens as $t) {
+                    $handler->destroy($t->last_session_id);
+                }
+                $this->loginTokenTable->deleteByUserId($cookie['user_id']);
+                $this->sendLoginTokenWarningEmail($user);
+                return null;
+            }
+        }
+        return $user;
+    }
+
+    /**
+     * Create a new login token
+     *
+     * @param \VuFind\Db\Row\User $user      user
+     * @param string              $series    login token series
+     * @param string              $sessionId Session identifier
+     * @param int                 $expires   Token expiration date
+     *
+     * @return void
+     */
+    public function createToken(\VuFind\Db\Row\User $user, string $series = '', string $sessionId = '', $expires = 0)
+    {
+        $token = bin2hex(random_bytes(32));
+        $series = $series ? $series : bin2hex(random_bytes(32));
+        $browser = '';
+        $platform = '';
+        try {
+            $userInfo = get_browser(null, true) ?? [];
+            $browser = $userInfo['browser'];
+            $platform = $userInfo['platform'];
+        } catch (\Exception $e) {
+            // Problem with browscap.ini, continue without
```

Comment:
Do we want to tolerate this situation? Would it be better to have a config setting to "opt in" to ignoring browscap.ini problems, so people don't accidentally turn on the feature in a less secure configuration than they expect?

Reply:
Not quite sure how to handle this. I'm not sure if there is actually a realistic scenario where this feature would be on without browser detection. We could do the opt in, but should we simply require a working browscap.ini? If we expect someone to use this without browser detection, should there be changes in the UI as well when no platform or browser is detected?

Reply:
I think requiring a working browscap.ini is reasonable. It makes sense to implement this as securely as possible, and that seems like an important layer to include. If somebody wants it without that feature, we can address the need when we have an understood use case for it.